File: Easy/Max Sum without Adjacents/max-sum-without-adjacents.py
#User function Template for python3
class Solution:
	
	def findMaxSum(self,arr, n):
		# code here
		if n==1 :
		    return arr[0]
		if n==2 :
		    return max(arr[0],arr[1])
        dp=[0]*(n+1)
        dp[0]=arr[0]
        dp[1]=max(arr[0],arr[1])
        for i in range(2,n) :
            dp[i]=max(dp[i-1],arr[i]+dp[i-2])
        return dp[n-1]

#{ 
 # Driver Code Starts
#Initial Template for Python 3




if __name__ == '__main__':
    tc = int(input())
    while tc > 0:
        n = int(input())
        arr = list(map(int, input().strip().split()))
        ob = Solution()
        ans = ob.findMaxSum(arr, n)
        print(ans)
        tc -= 1

# } Driver Code Ends

File: Easy/Max Sum without Adjacents/max-sum-without-adjacents.java
//{ Driver Code Starts
//Initial Template for Java



import java.util.*;
import java.io.*;

public class Main {

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int tc = Integer.parseInt(br.readLine().trim());
        while (tc-- > 0) {
            String[] inputLine;
            int n = Integer.parseInt(br.readLine().trim());
            int[] arr = new int[n];
            inputLine = br.readLine().trim().split(" ");
            for (int i = 0; i < n; i++) {
                arr[i] = Integer.parseInt(inputLine[i]);
            }

            int ans = new Solution().findMaxSum(arr, n);
            System.out.println(ans);
        }
    }
}
// } Driver Code Ends


//User function Template for Java

class Solution {
    int solve(int i,int n,int arr[],int dp[])
    {
        if (i>=n)
        {
            return 0;
        }
        if (dp[i]!=0)
        {
            return dp[i];
        }
        int tk=arr[i]+solve(i+2,n,arr,dp);
        int ntk=solve(i+1,n,arr,dp);
        dp[i]=Math.max(tk,ntk);
        return dp[i];
    }
    int findMaxSum(int arr[], int n) {
        // code here
        int dp[]=new int[n+1];
        return solve(0,n,arr,dp);
    }
    
}